doc/design/zarr-support-3.md

```diff
@@ -0,0 +1,149 @@
+# Zarr Support
+A zarr archive contains directories and zarr files. 
+Zarr archives are stored in their entirety in S3.
+Zarr archives are represented by a single Asset through the API.
+A tree hash is computed as the zarr archive is uploaded.
+The tree hash details (all nodes) are stored on S3 to avoid bloating the DB.
+The intended use of Zarr archives is primarily to support use cases where data are too big to be copied or to be stored in a single file due to available storage limits. However, Zarr archives can be small as well.
+
+# Requirements
+1. Zarr archives are stored in a "directory" in S3.
+1. Each zarr archive corresponds to a single Asset.
+1. The CLI uses some kind of tree hashing scheme to compute a checksum for the entire zarr archive.
+1. The API verifies the checksum immediately after upload, so the validation status is available in the response to the upload REST request.
+1. The system can theoretically handle zarr files with ~1 million subfiles, each of size 64 * 64 * 64 bytes ~= 262 kilobytes.
+1. Zarr metadata must be mutable for zarr files in draft dandisets.
+1. Zarr archives must not be copied, as they are too large to reasonably store multiple copies.
+1. Zarr archives are immutable as long as they are contained by a published dandiset.
+
+# Implementation
+
+## API endpoints
+* **POST /api/zarr/**
+
+  Create a new zarr archive.
+
+  Returns a zarr ID
+
+* **GET /api/zarr/{zarr_id}/**
+
+  Returns some information about the zarr archive, like name, S3 path+url, and checksum
+
+* **GET /api/zarr/{zarr_id}/{path}**
+
+  When the path exists in S3, returns some information about the zarr file, like name, S3 path+url, and checksum.
+  When the path is a directory in S3, returns a paginated list of files and directories. 
+  Returns 404 if the path is not present in S3.
+
+* **GET /api/zarr/{zarr_id}/upload/**
+
+  Returns a 204 if an upload is in progress, or a 404 if an upload is not in progress.
+
+* **POST /api/zarr/{zarr_id}/upload/**
+
+  Ask to upload a batch of zarr files as part of the zarr archive.
+  No more batch uploads are permitted until this one is completed or aborted.
+
+  Requires a list of file paths and ETags (md5 checksums).
+  The number of files being uploaded must be less than some experimentally defined limit (say ~500).
```

Comment:
should we provision some return error code to instruct client to "try less" if for some reason `500` would still be "too many"?

Reply:
```suggestion
  The number of files being uploaded must be less than some experimentally defined limit (say ~500).
  If the limit is exceeded, return response 400.
```